precovery/orbit.py

```diff
@@ -0,0 +1,443 @@
+            dynmodel = "N"
+        elif method == PropagationIntegrator.TWO_BODY:
+            dynmodel = "2"
+        else:
+            raise ValueError("unexpected propagation method %r" % method)
+
+        eph, err = pyoorb.pyoorb.oorb_ephemeris_basic(
+            in_orbits=self._state_vector,
+            in_obscode=obscode,
+            in_date_ephems=epochs_array,
+            in_dynmodel=dynmodel,
+        )
+        # print(epochs_array.shape)
+        # print(eph.shape)
+        assert err == 0, "There was an issue with the pyoorb ephemeris generation"
+        return [Ephemeris(eph[0, i, :]) for i in range(epochs_array.shape[0])]
+
+    def precover_remote(
+        self,
+        tolerance: float = 30 * ARCSEC,
+        max_matches: Optional[int] = None,
+        start_mjd: Optional[float] = None,
+        end_mjd: Optional[float] = None,
+        window_size: Optional[float] = None,
+    ):
+        """
+        Find observations which match orbit in the database. Observations are
+        searched in descending order by mjd.
+
+        Expects three environment variables:
+        PRECOVERY_API_SINGLEORBIT_URL
+        PRECOVERY_API_SINGLEORBIT_USERNAME
+        PRECOVERY_API_SINGLEORBIT_PASSWORD
+
+        max_matches: End once this many matches have been found. If None, find
+        all matches.
+
+        start_mjd: Only consider observations from after this epoch
+        (inclusive). If None, find all.
+
+        end_mjd: Only consider observations from before this epoch (inclusive).
+        If None, find all.
+
+        window_size: UNIMPLEMENTED
+        """
+        # basically:
+        """
+        find all windows between start and end of given size
+        for each window:
+            propagate to window center
+            for each unique epoch,obscode in window:
+                propagate to epoch
+                find frames which match healpix of propagation
+                for each matching frame
+                    find matching observations
+                    for each matching observation
+                        yield match
+        """
+
+        # Check for environment set vars
+        precovery_singleorbit_url = getenv("PRECOVERY_API_SINGLEORBIT_URL")
+        api_username = getenv("PRECOVERY_API_SINGLEORBIT_USERNAME")
+        # Note Password suffix results in encryption and storage in systems keys for conda env
+        # (https://anaconda-project.readthedocs.io/en/latest/user-guide/tasks/work-with-variables.html#adding-an-encrypted-variable)
+        api_password = getenv("PRECOVERY_API_SINGLEORBIT_PASSWORD")
+
+        if not (precovery_singleorbit_url and api_username and api_password):
+            raise ValueError(
+                """one of required environment variables unset, expecting PRECOVERY_API_SINGLEORBIT_URL,
+            PRECOVERY_API_SINGLEORBIT_USERNAME, PRECOVERY_API_SINGLEORBIT_PASSWORD"""
+            )
+
+        if self._orbit_type == OrbitElementType.KEPLERIAN:
```

Comment:
Is it a good or bad idea to choose an orbit type to store internally on initialization? We could store the original state vector and type and then always normalize on a single representation? Obviously bad idea if it introduces error or uncertainty.